### ISSUE TITLE
setting Resources.MemorySwap unlimited after type change is messy

### ISSUE THREAD

**Issue by sjenning** [opencontainers/runc#1421]
https://github.com/opencontainers/runc/pull/1375 changed `Resources.MemorySwap` to a `uint64`.

https://github.com/opencontainers/runc/blob/master/libcontainer/configs/cgroup_unix.go#L53-L54

Comment on that field is set to `-1` for unlimited, which can't be done with a `uint64`

This is blocking a vendored dep bump in kube: https://github.com/kubernetes/kubernetes/issues/43856

@derekwaynecarr

**Comment by sjenning** [comment]
ok, well... I found this https://github.com/opencontainers/runc/blob/master/libcontainer/cgroups/fs/memory.go#L107-L116

Apparently the solution is `uint64(-1)` which seems messy

**Comment by sjenning** [comment]
Even more messy,  `uint64(-1)` doesn't even compile.  You have to trick it with:

```
unlimited := -1
res.MemorySwap = uint64(unlimited)
```

**Comment by derekwaynecarr** [comment]
@mrunalp @vishh -- i find this awkward to say the least.

**Comment by dqminh** [comment]
Yah i agree its a bit awkward, `math.MaxUint64` should do the job as well though.

**Comment by justincormack** [comment]
The kernel specifically accepts `-1` in its parser, using `page_counter_memparse(args, "-1",...)` but we no longer allow this since this change, so it no longer matches the kernel behaviour, even if it is a "sort of" `uint64` it is a special one that also accepts `-1`.

In addition the maximum meaningful value is 2^63 due to the kernel/user memory split.

In addition some older kernels (RHEL 7.2 at least) will set the limits to `0` if they are larger than 2^63, and then immediately crash, so setting them to `2^64-1` makes them crash, while the previous `-1` worked correctly.

So we should revert this whole change.

cc @tiborvass

**Comment by hqhq** [comment]
@justincormack `uint64` is actually the type kernel uses for memory limit, it accepts `-1` because we take this value as string in cgroup API, we can do the same thing in our user space tools if we see fit, or we can force change `-1` to uint64 which also works.

> In addition the maximum meaningful value is 2^63 due to the kernel/user memory split.
>
> In addition some older kernels (RHEL 7.2 at least) will set the limits to 0 if they are larger than 2^63, and then immediately crash, so setting them to 2^64-1 makes them crash, while the previous -1 worked correctly.

These are points I didn't realize, I'll double check these, doesn't make a lot of sense to me though.

**Comment by hqhq** [comment]
@justincormack I realized that on old kernels (before 3.11), the max resource in kernel is LLONG_MAX, which is wrong, and it's fixed by: http://marc.info/?l=linux-mm&m=137541397220314&w=2

So on RHEL 7.2, you just can't write values larger than 2^63, this is some compatibility issues we should fix, I just opened a PR #1494 to fix this.

I still think use uint64 for memory entries to keep consistency with Linux kernel is the right approach.

WDYT?
